components/webext-storage/ffi/src/error_codes.rs

```diff
@@ -0,0 +1,22 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/// An unexpected error occurred which likely cannot be meaningfully handled
```

Comment:
The pattern we've been using for our other FFIs is a bit different—we have an `ffi` module inside our parent crate, and that has an `error_codes` submodule and conversions between that crate's errors and `ExternError`. How do you all feel about changing that slightly?

My thoughts were:

1. I'm not a big fan of submodules in general (except for `mod tests`), because it's harder to see when you're looking at the directory structure. Places in particular is a bit of a tangle: there's a top-level `bookmark_sync` module, _and_ a `mod bookmark_sync` declaration inside `storage/bookmarks.rs` (ditto for history), but you wouldn't guess that from looking at the file names. When working on that code, I get a little disoriented at first—"wait, `bookmarks` doesn't have a `bookmark_sync`, where's that coming from...oh." That said, `ffi::error_codes` is only one level of nesting, and not that bad.
2. But, we have an `ffi` module and an `ffi` "subcrate", and I do get those confused as well.
3. Having the `implement_into_ffi_by_{json, protobuf}` far away from where the types are declared gets a little confusing, too, but maybe it's my mental model that's off (I go for the type first, but grouping those into `ffi.rs` feels like a "concept first").

But I'd be happy to change this to follow the others if this is harder/clunkier/we're just not feeling it!

Reply:
+1 from me - flat is better than nested. I agree places is too nested (and that's my fault IIRC)

Reply:
Nah, I definitely share that blame, too! 😆 A year and a half ago, we were all still figuring out good patterns for Rust!

Reply:
Nested in a separate file feels worse to me, I'd rather these just be constants in the parent maybe -- It's hard for me to imagine a time when you edit this file but don't also edit the other, but honestly I don't care much and it seems I'm the only one that feels this way.